packages/teleterm/src/ui/services/clusters/clustersService.ts

```diff
@@ -58,7 +59,8 @@ export class ClustersService extends ImmutableStore<ClustersServiceState> {
   constructor(
     public client: tsh.TshClient,
     private mainProcessClient: MainProcessClient,
-    private notificationsService: NotificationsService
+    private notificationsService: NotificationsService,
+    private usageReporterService: UsageEventService
```

Comment:
Adding `usageReporterService` to each place that needs to report an event is not necessarily a bad approach, but did you consider other approaches?

Have you heard of aspect-oriented programming?

* https://wiki.c2.com/?AspectOrientedProgramming
* https://thecodebarbarian.com/practical-aspect-oriented-programming-in-javascript.html

It's something to consider once we get around to adding more and more usage metrics. Sending those usage events seems like a very cross-cutting concern so we could be better off using AOP-like approach or just having those different services fire off events to a single event emitter and then maybe sending actual usage events based on those.

I'm just putting it out there, for now passing `usageReporterService` around should be alright.

Reply:
I think I have never heard the name :) 
So in our case it would mean intercepting calls to `tshd`? (If no, I'd appreciate some small example)

Reply:
> So in our case it would mean intercepting calls to `tshd`? (If no, I'd appreciate some small example)

Yeah, or to `ClustersService`. It wouldn't even need to follow the principles of AOP, I was thinking of any solution where the object performing the action that needs to be reported doesn't need to depend on the usage reporting service directly.

This would also solve the problem with the circular dependencies between `ClustersService` and `UsageEventService`.

Reply:
I was thinking about splitting `UsageEventService` into two services (one would capture events and the second would report them to `tshd` and add cluster properties), but it complicated things even more, as the "capturer" had to be initalized with "reporter" using a method like (`capturer.init(reporter)`).
I left a TODO about removing the circular dependency issue.